### ISSUE TITLE
ok_or_else suggested over ok_or with tuple struct type

### ISSUE THREAD

**Issue by Eddio0141** [rust-lang/rust-clippy#8525]
### Summary

Clippy gave me a warning on using `ok_or` contaning a tuple struct, instead wanting me to replace it with `ok_or_else` containing a closure with the tuple struct inside it.
This happens the same with an enum variant that's a tuple.

### Lint Name

or_fun_call

### Reproducer

basic code to reproduce warning:
```rust
struct ErrorStruct(String);

let some_option = Some(0);
let _result = some_option.ok_or(ErrorStruct("something went wrong".to_string()));
```

which gives me a warning:
`warning: use of ok_or followed by a function call`
and the suggestion:
`help: try this: ok_or_else(|| ErrorStruct("something went wrong".to_string()))`

### Version

```text
rustc 1.59.0 (9d1b2106e 2022-02-23)
binary: rustc
commit-hash: 9d1b2106e23b1abd32fce1f17267604a5102f57a
commit-date: 2022-02-23
host: x86_64-pc-windows-msvc
release: 1.59.0
LLVM version: 13.0.0
```

**Comment by PatchMixolydic** [comment]
This seems correct; the original code will allocate a `String` for `_result`'s error message even when `some_option` is `Some(_)`. The suggested code will only call the closure and allocate the `String` if `some_option` is `None` (which is especially good for the reproducer since `some_option` is never `None`).

**Comment by Eddio0141** [comment]
Ah ok, thanks for letting me know. Sorry for the false issue.
